sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -329,6 +329,22 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
       Row(3, "c") ::
       Row(4, "d") :: Nil)
     checkAnswer(lowerCaseData.intersect(upperCaseData), Nil)
+
+    // check null equality
+    checkAnswer(
+      nullInts.intersect(nullInts),
+      Row(1) ::
+      Row(2) ::
+      Row(3) ::
+      Row(null) :: Nil)
+
+    // check duplicate values
+    checkAnswer(
+      allNulls.intersect(allNulls),
+      Row(null) ::
+      Row(null) ::
+      Row(null) ::
+      Row(null) :: Nil)
```

Comment:
Is the expected answer right?

 Also, can you also add the test mentioned by @nongli (handling null is kind of special. So let's also have a regular case).